contracts/references/SampleTokenPaymaster.sol

```diff
@@ -0,0 +1,217 @@
+    /// @param _entryPoint The EntryPoint contract used in the Account Abstraction infrastructure.
+    /// @param _wrappedNative The ERC-20 token that wraps the native asset for current chain.
+    /// @param _uniswap The Uniswap V3 SwapRouter contract.
+    /// @param _tokenPaymasterConfig The configuration for the Token Paymaster.
+    /// @param _oracleHelperConfig The configuration for the Oracle Helper.
+    /// @param _uniswapHelperConfig The configuration for the Uniswap Helper.
+    /// @param _owner The address that will be set as the owner of the contract.
+    constructor(
+        IERC20Metadata _token,
+        IEntryPoint _entryPoint,
+        IERC20 _wrappedNative,
+        ISwapRouter _uniswap,
+        TokenPaymasterConfig memory _tokenPaymasterConfig,
+        OracleHelperConfig memory _oracleHelperConfig,
+        UniswapHelperConfig memory _uniswapHelperConfig,
+        address _owner
+    )
+    BasePaymaster(
+    _entryPoint
+    )
+    OracleHelper(
+    _oracleHelperConfig
+    )
+    UniswapHelper(
+    _token,
+    _wrappedNative,
+    _uniswap,
+    _uniswapHelperConfig
+    )
+    {
+        setTokenPaymasterConfig(_tokenPaymasterConfig);
+        transferOwnership(_owner);
+    }
+
+    /// @notice Updates the configuration for the Token Paymaster.
+    /// @param _tokenPaymasterConfig The new configuration struct.
+    function setTokenPaymasterConfig(
+        TokenPaymasterConfig memory _tokenPaymasterConfig
+    ) public onlyOwner {
+        require(_tokenPaymasterConfig.priceMarkup <= 2 * PRICE_DENOMINATOR, "TPM: price markup too high");
+        require(_tokenPaymasterConfig.priceMarkup >= PRICE_DENOMINATOR, "TPM: price markup too low");
+        tokenPaymasterConfig = _tokenPaymasterConfig;
+        emit ConfigUpdated(_tokenPaymasterConfig);
+    }
+
+    function setUniswapConfiguration(
+        UniswapHelperConfig memory _uniswapHelperConfig
+    ) external onlyOwner {
+        _setUniswapHelperConfiguration(_uniswapHelperConfig);
+    }
+
+    /// @notice Allows the contract owner to withdraw a specified amount of tokens from the contract.
+    /// @param to The address to transfer the tokens to.
+    /// @param amount The amount of tokens to transfer.
+    function withdrawToken(address to, uint256 amount) external onlyOwner {
+        SafeERC20.safeTransfer(token, to, amount);
+    }
+
+    /// @notice Validates a paymaster user operation and calculates the required token amount for the transaction.
+    /// @param userOp The user operation data.
+    /// @param requiredPreFund The maximum cost (in native token) the paymaster has to prefund.
+    /// @return context The context containing the token amount and user sender address (if applicable).
+    /// @return validationResult A uint256 value indicating the result of the validation (always 0 in this implementation).
+    function _validatePaymasterUserOp(PackedUserOperation calldata userOp, bytes32, uint256 requiredPreFund)
```

Comment:
ok, let's discuss here. @ShivaanshK 

**Q. Does unaccountedGas need to be used at all in calculating precharge?**

> the maxCost that is being passed into validatePaymasterUserOp is calculated using the limits and the userop can’t use more gas than that

yes, true. it can only deviate by 10% otherwise there is penalty. so requiredPrefund would have paymasterPostOpGasLimit component? (todo: cross verify)
then this is really on how well paymaster service does it's job. 

and in _postOp you will get actualGasCost. although diff may be very low in requiredPrefund and actualGasCost there is no harm in keeping unaccountedGasCost [ we should find out this value being set and used in live paymaster and see how off is it / or in our test case ] although we might have to keep it on token config basis but common global can't hurt.

then we shall keep these below lines too

```
uint256 refundPostopCost = tokenPaymasterConfig.refundPostopCost;
require(refundPostopCost < userOp.unpackPostOpGasLimit(), "TPM: postOpGasLimit too low");
```

Reply:
"My intuition is that even if postOp revert due to exceeding the gas limit, the paymaster still needs to pay the gas used during postOp"
also didn't get this @ShivaanshK